File: python/sigdet/stats_sigdet_ROC.py
#!/usr/bin/env python
# coding: utf-8

# # COURSE: Master statistics and machine learning: Intuition, Math, code
# ##### COURSE URL: udemy.com/course/statsml_x/?couponCode=202006 
# ## SECTION: Signal detection theory
# ### VIDEO: ROC curves
# #### TEACHER: Mike X Cohen, sincxpress.com

# In[ ]:


import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as stats


# In[ ]:


## first, re-create the dp and rb matrices from previous lectures

x  = np.arange(.01,1,.01)
dp = np.tile(stats.norm.ppf(x),(99,1)).T - np.tile(stats.norm.ppf(x),(99,1))
rb = -( np.tile(stats.norm.ppf(x),(99,1)).T + np.tile(stats.norm.ppf(x),(99,1)) )/2


# In[ ]:


## create the 2D bias spaces and plot their ROC curves

rb2plot = [.3, .5, .9, 1.5] # d'/bias levels 
tol = .01 # tolerance for matching levels
colorz = 'rbmk'

# setup the figure
fig,ax = plt.subplots(1,2,figsize=(10,5))

# show the 2D spaces
ax[0].imshow(dp,extent=[x[0],x[-1],x[0],x[-1]],origin='lower')
ax[0].set_xlabel('False alarm rate')
ax[0].set_ylabel('Hit rate')
ax[0].set_title("d'")

ax[1].imshow(rb,extent=[x[0],x[-1],x[0],x[-1]],origin='lower')
ax[1].set_xlabel('False alarm rate')
ax[1].set_ylabel('Hit rate')
ax[1].set_title('Response bias')



### now draw the isocontours
for i in range(len(rb2plot)):
    
    # find d' points
    idx = np.where((dp>rb2plot[i]-tol) & (dp<rb2plot[i]+tol))
    ax[0].plot(x[idx[1]],x[idx[0]],'o-',color=colorz[i])
    
    # find bias points
    idx = np.where((rb>rb2plot[i]-tol) & (rb<rb2plot[i]+tol))
    ax[1].plot(x[idx[1]],x[idx[0]],'o-',color=colorz[i])


plt.show()

File: MATLAB/sigdet/stats_sigdet_ROC.m
%%
%   COURSE: Master statistics and machine learning: intuition, math, code										
%      URL: udemy.com/course/statsml_x/?couponCode=202304 
% 
%  SECTION: Signal detection theory
%    VIDEO: ROC curves
% 
%  TEACHER: Mike X Cohen, sincxpress.com
%

%%

% a clear MATLAB workspace is a clear mental workspace
close all; clear; clc

%% recreate the matrices

x  = .01:.01:.99;
dp = norminv(x)' - norminv(x);
rb = -( norminv(x)' + norminv(x) )/2;

%% create the 2D spaces

figure(1), clf
subplot(121), hold on
contourf(x,x,dp,80,'linecolor','none')
xlabel('False alarm rate')
ylabel('Hit rate')
title('d''')
axis square

subplot(122), hold on
contourf(x,x,rb,80,'linecolor','none')
xlabel('False alarm rate')
ylabel('Hit rate')
title('Response bias')
axis square

%% now draw the isocontours

colorz = 'rbmk';
rb2plot = [.3 .5 .9 1.5];
tol = .01;

for dpi=1:length(rb2plot)
    
    % d': find points and plot isosensitivity curves
    idx = find(dp>rb2plot(dpi)-tol & dp<rb2plot(dpi)+tol);
    [yi,xi] = ind2sub(size(dp),idx);
    subplot(121)
    plot(x(xi),x(yi),[ colorz(dpi) 'o-' ],'markersize',4,'markerfacecolor',colorz(dpi))
    
    % repeat for response bias
    idx = find(rb>rb2plot(dpi)-tol & rb<rb2plot(dpi)+tol);
    [yi,xi] = ind2sub(size(rb),idx);
    subplot(122)
    plot(x(xi),x(yi),[ colorz(dpi) 'o-' ],'markersize',4,'markerfacecolor',colorz(dpi))
end

%% done.
